packages/compiler/src/render3/view/t2_binder.ts

```diff
@@ -495,18 +522,38 @@ class TemplateBinder extends RecursiveAstVisitor implements Visitor {
     return {expressions, symbols, nestingLevel, usedPipes, eagerPipes, deferBlocks};
   }
 
-  private ingest(template: Template|Node[]): void {
-    if (template instanceof Template) {
+  private ingest(nodeOrNodes: ScopedNode|Node[]): void {
+    if (nodeOrNodes instanceof Template) {
       // For <ng-template>s, process only variables and child nodes. Inputs, outputs, templateAttrs,
       // and references were all processed in the scope of the containing template.
-      template.variables.forEach(this.visitNode);
-      template.children.forEach(this.visitNode);
+      nodeOrNodes.variables.forEach(this.visitNode);
+      nodeOrNodes.children.forEach(this.visitNode);
 
       // Set the nesting level.
-      this.nestingLevel.set(template, this.level);
+      this.nestingLevel.set(nodeOrNodes, this.level);
+    } else if (nodeOrNodes instanceof IfBlockBranch) {
+      // TODO(crisbeto): uncomment this when rebasing #51690.
+      // if (nodeOrNodes.expressionAlias !== null) {
+      //   this.visitNode(nodeOrNodes.expressionAlias);
+      // }
+      nodeOrNodes.children.forEach(this.visitNode);
+      this.nestingLevel.set(nodeOrNodes, this.level);
+    } else if (nodeOrNodes instanceof ForLoopBlock) {
+      // TODO(crisbeto): uncomment this when rebasing #51690.
+      // this.visitNode(nodeOrNodes.item);
+      // Object.values(nodeOrNodes.contextVariables).forEach(v => this.visitNode(v));
+      nodeOrNodes.trackBy.visit(this);
+      nodeOrNodes.children.forEach(this.visitNode);
+      this.nestingLevel.set(nodeOrNodes, this.level);
+    } else if (
+        nodeOrNodes instanceof DeferredBlock || nodeOrNodes instanceof DeferredBlockError ||
+        nodeOrNodes instanceof DeferredBlockPlaceholder ||
+        nodeOrNodes instanceof DeferredBlockLoading) {
```

Comment:
Since we use this check in a couple places, it'd be great to extract it to a separate fn:
```suggestion
    } else if (isScopedNode(nodeOrNodes)) {
```
We can check all node types there (including `IfBlockBranch` and `ForLoopBlock`), so we can do do type narrowing too:
```
function isScopedNode(node: unknown): node is ScopedNode { ... }
```

Reply:
I agree that this felt a bit weird when I was writing it, but there are a couple of reasons why I think we should keep it:
1. Once template type checking is implemented for control flow, there will be different handling for each `ScopedNode`, with the exception of the deferred ones.
2. Using the `if`/`else if` like we do here will result in an error whenever we add a node type to `ScopedNode`, which is good because it makes it easier to see where we need to adjust the logic. In comparison, if we do something like following we won't get such an error:

```
if (!isScopedNode(node)) {
  return ...;
}

if (node instanceof IfBlock) {
   ...
} else if (node instanceof DeferredBlock) {
   ...
}
```